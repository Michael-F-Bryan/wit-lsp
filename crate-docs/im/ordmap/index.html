<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="An ordered map."><title>im::ordmap - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/SourceSerif4-Regular-46f98efaafac5295.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/FiraSans-Regular-018c141bf0843ffd.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/FiraSans-Medium-8f9a781e4970d388.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2"><link rel="stylesheet" href="../../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../../static.files/rustdoc-5bc39a1768837dd0.css"><meta name="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="im" data-themes="" data-resource-suffix="" data-rustdoc-version="1.77.2 (25ef9e3d8 2024-04-09)" data-channel="1.77.2" data-search-js="search-dd67cee4cfa65049.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../../static.files/storage-4c98445ec4002617.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../static.files/main-48f368f3872407c8.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-04d5337699b92874.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-16x16-8b506e7a72182f1c.png"><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../im/index.html">im</a><span class="version">15.1.0</span></h2></div><h2 class="location"><a href="#">Module ordmap</a></h2><div class="sidebar-elems"><section><ul class="block"><li><a href="#macros">Macros</a></li><li><a href="#structs">Structs</a></li><li><a href="#enums">Enums</a></li></ul></section><h2><a href="../index.html">In crate im</a></h2></div></nav><div class="sidebar-resizer"></div>
    <main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><div id="sidebar-button" tabindex="-1"><a href="../../im/all.html" title="show sidebar"></a></div><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" tabindex="-1"><a href="../../help.html" title="help">?</a></div><div id="settings-menu" tabindex="-1"><a href="../../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../../static.files/wheel-7b819b6101059cd0.svg"></a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1>Module <a href="../index.html">im</a>::<wbr><a class="mod" href="#">ordmap</a><button id="copy-path" title="Copy item path to clipboard"><img src="../../static.files/clipboard-7571035ce49a181d.svg" width="19" height="18" alt="Copy item path"></button></h1><span class="out-of-band"><a class="src" href="../../src/im/ord/map.rs.html#5-2649">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>An ordered map.</p>
<p>An immutable ordered map implemented as a [B-tree] <a href="https://en.wikipedia.org/wiki/B-tree">1</a>.</p>
<p>Most operations on this type of map are O(log n). A
<a href="../hashmap/struct.HashMap.html"><code>HashMap</code></a> is usually a better choice for
performance, but the <code>OrdMap</code> has the advantage of only requiring
an <a href="https://doc.rust-lang.org/std/cmp/trait.Ord.html"><code>Ord</code></a> constraint on the key, and of being
ordered, so that keys always come out from lowest to highest,
where a <a href="../hashmap/struct.HashMap.html"><code>HashMap</code></a> has no guaranteed ordering.</p>
</div></details><h2 id="macros" class="section-header">Macros<a href="#macros" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="macro" href="macro.ordmap.html" title="macro im::ordmap::ordmap">ordmap</a></div><div class="desc docblock-short">Construct a map from a sequence of key/value pairs.</div></li></ul><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="struct" href="struct.ConsumingIter.html" title="struct im::ordmap::ConsumingIter">ConsumingIter</a></div><div class="desc docblock-short">A consuming iterator over an ordered set.</div></li><li><div class="item-name"><a class="struct" href="struct.DiffIter.html" title="struct im::ordmap::DiffIter">DiffIter</a></div><div class="desc docblock-short">An iterator over the differences between two maps.</div></li><li><div class="item-name"><a class="struct" href="struct.Iter.html" title="struct im::ordmap::Iter">Iter</a></div><div class="desc docblock-short">An iterator over the key/value pairs of a map.</div></li><li><div class="item-name"><a class="struct" href="struct.Keys.html" title="struct im::ordmap::Keys">Keys</a></div><div class="desc docblock-short">An iterator ove the keys of a map.</div></li><li><div class="item-name"><a class="struct" href="struct.NodeDiffIter.html" title="struct im::ordmap::NodeDiffIter">NodeDiffIter</a></div><div class="desc docblock-short">An iterator over the differences between two ordered sets.</div></li><li><div class="item-name"><a class="struct" href="struct.OccupiedEntry.html" title="struct im::ordmap::OccupiedEntry">OccupiedEntry</a></div><div class="desc docblock-short">An entry for a mapping that already exists in the map.</div></li><li><div class="item-name"><a class="struct" href="struct.OrdMap.html" title="struct im::ordmap::OrdMap">OrdMap</a></div><div class="desc docblock-short">An ordered map.</div></li><li><div class="item-name"><a class="struct" href="struct.OrdMapPool.html" title="struct im::ordmap::OrdMapPool">OrdMapPool</a></div><div class="desc docblock-short">A memory pool for the appropriate node type.</div></li><li><div class="item-name"><a class="struct" href="struct.RangedIter.html" title="struct im::ordmap::RangedIter">RangedIter</a></div><div class="desc docblock-short">An iterator over an ordered set.</div></li><li><div class="item-name"><a class="struct" href="struct.VacantEntry.html" title="struct im::ordmap::VacantEntry">VacantEntry</a></div><div class="desc docblock-short">An entry for a mapping that does not already exist in the map.</div></li><li><div class="item-name"><a class="struct" href="struct.Values.html" title="struct im::ordmap::Values">Values</a></div><div class="desc docblock-short">An iterator over the values of a map.</div></li></ul><h2 id="enums" class="section-header">Enums<a href="#enums" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="enum" href="enum.DiffItem.html" title="enum im::ordmap::DiffItem">DiffItem</a></div><div class="desc docblock-short">A description of a difference between two ordered maps.</div></li><li><div class="item-name"><a class="enum" href="enum.Entry.html" title="enum im::ordmap::Entry">Entry</a></div><div class="desc docblock-short">A handle for a key and its associated value.</div></li><li><div class="item-name"><a class="enum" href="enum.NodeDiffItem.html" title="enum im::ordmap::NodeDiffItem">NodeDiffItem</a></div><div class="desc docblock-short">A description of a difference between two ordered sets.</div></li></ul></section></div></main></body></html>