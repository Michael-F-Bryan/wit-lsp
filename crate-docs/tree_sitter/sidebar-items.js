window.SIDEBAR_ITEMS = {"constant":["ARRAY_HEADER","LANGUAGE_VERSION","MIN_COMPATIBLE_LANGUAGE_VERSION","PARSER_HEADER"],"enum":["CaptureQuantifier","LogType","QueryErrorKind","QueryPredicateArg"],"fn":["set_allocator","wasm_stdlib_symbols"],"mod":["ffi"],"struct":["IncludedRangesError","InputEdit","Language","LanguageError","LanguageRef","LookaheadIterator","LossyUtf8","Node","Parser","Point","Query","QueryCapture","QueryCaptures","QueryCursor","QueryError","QueryMatch","QueryMatches","QueryPredicate","QueryProperty","Range","Tree","TreeCursor"],"trait":["TextProvider"]};