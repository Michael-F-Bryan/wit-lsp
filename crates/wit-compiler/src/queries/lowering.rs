use im::{OrdMap, Vector};
use tree_sitter::Node;

use crate::{
    ast::{self, AstNode, HasAttr, HasIdent},
    diagnostics::{Diagnostic, Diagnostics, Location},
    hir,
    queries::{SourceFile, Workspace},
    Db, Text,
};

/// Parse a file and lower it from its [`crate::ast`] representation to the
/// [`crate::hir`] representation.
///
/// Nodes that contain syntactic or semantic errors will be ignored and a
/// corresponding [`Diagnostic`] will be emitted.
#[salsa::tracked]
#[tracing::instrument(level = "debug", skip_all, fields(file = %file.path(db)))]
pub fn lower(db: &dyn Db, _ws: Workspace, file: SourceFile) -> hir::Package {
    let ast = crate::queries::parse(db, file);
    let src = file.contents(db);
    let root = ast.source_file(db);

    let mut ctx = State::new(db, src, file);
    ctx.process_file(root);

    ctx.finish()
}

/// A big bundle of state used while lowering.
struct State<'a> {
    db: &'a dyn Db,
    src: &'a str,
    file: SourceFile,
    decl: Option<hir::PackageDeclaration>,
    interfaces: OrdMap<Text, hir::Interface>,
    worlds: OrdMap<Text, hir::World>,
    names: OrdMap<Text, NamedItem<'a>>,
}

impl<'a> State<'a> {
    pub fn new(db: &'a dyn Db, src: &'a str, file: SourceFile) -> Self {
        State {
            db,
            file,
            decl: None,
            interfaces: OrdMap::default(),
            names: OrdMap::new(),
            src,
            worlds: OrdMap::default(),
        }
    }

    fn process_file(&mut self, root: ast::SourceFile<'a>) {
        self.decl = self.lower_package_decl(root);

        for item in root.iter_top_level_items().flat_map(named_item) {
            let name = item.ident().value(self.src);

            match item {
                NamedItem::World(node) => {
                    if let Some(world) = self.lower_world(node) {
                        self.push_world(name, world, node);
                    }
                }
                NamedItem::Interface(node) => {
                    if let Some(world) = self.lower_interface(node) {
                        self.push_interface(name, world, node);
                    }
                }
                NamedItem::Use(_) => todo!(),
            }
        }
    }

    fn push_interface(
        &mut self,
        name: &str,
        interface: hir::Interface,
        node: ast::InterfaceItem<'a>,
    ) {
        let name = Text::from(name);

        if self.insert_named(self.db, name.clone(), node.into()) {
            self.interfaces.insert(name, interface);
        }
    }

    fn push_world(&mut self, name: &str, world: hir::World, node: ast::WorldItem<'a>) {
        let name = Text::from(name);

        if self.insert_named(self.db, name.clone(), node.into()) {
            self.worlds.insert(name, world);
        }
    }

    fn insert_named(&mut self, db: &dyn Db, name: Text, node: NamedItem<'a>) -> bool {
        match self.names.entry(name) {
            im::ordmap::Entry::Occupied(entry) => {
                let name = entry.key().clone();
                let original_node = *entry.get();
                drop(entry);
                let location = self.location(db, node.node());
                let original_definition = self.location(db, original_node.node());

                let diag = Diagnostic::duplicate_name(name, location, original_definition);
                Diagnostics::push(db, diag);
                false
            }
            im::ordmap::Entry::Vacant(entry) => {
                entry.insert(node);
                true
            }
        }
    }

    fn location(&self, db: &dyn Db, node: Node<'_>) -> Location {
        Location::new(self.file.path(db), node.range())
    }

    fn finish(self) -> hir::Package {
        let State {
            interfaces,
            worlds,
            decl,
            ..
        } = self;

        hir::Package {
            decl,
            interfaces: interfaces.into_iter().map(|(_, i)| i).collect(),
            worlds: worlds.into_iter().map(|(_, w)| w).collect(),
        }
    }

    fn lower_world(&mut self, world: ast::WorldItem<'_>) -> Option<hir::World> {
        let docs = world.docs(self.src);
        let name = world.identifier(self.src)?.into();
        let items = world
            .iter_items()
            .filter_map(|item| self.lower_world_item(item))
            .collect();

        Some(hir::World { name, docs, items })
    }

    fn lower_world_item(&mut self, item: ast::WorldItems<'_>) -> Option<hir::WorldItem> {
        if let Some(export) = item.export_item() {
            self.lower_export_item(export).map(hir::WorldItem::Export)
        } else if let Some(import) = item.import_item() {
            self.lower_import_item(import).map(hir::WorldItem::Import)
        } else if let Some(include) = item.include_item() {
            self.lower_include_item(include)
                .map(hir::WorldItem::Include)
        } else {
            None
        }
    }

    fn lower_include_item(&mut self, item: ast::IncludeItem<'_>) -> Option<hir::Include> {
        let _path = item.path()?;

        todo!()
    }

    fn lower_import_item(&mut self, item: ast::ImportItem<'_>) -> Option<hir::ExposableItem> {
        dbg!(item.syntax().to_sexp());

        if let Some(_path) = item.imported_item() {
            todo!()
        } else if let Some(_item) = item.imported_path() {
            todo!()
        } else {
            None
        }
    }

    fn lower_export_item(&mut self, item: ast::ExportItem<'_>) -> Option<hir::ExposableItem> {
        if let Some(_path) = item.exported_item() {
            todo!()
        } else if let Some(_item) = item.exported_item() {
            todo!()
        } else {
            None
        }
    }

    fn lower_interface(&mut self, interface: ast::InterfaceItem<'_>) -> Option<hir::Interface> {
        let docs = interface.docs(self.src);
        let name = interface.identifier(self.src)?.into();

        let items = interface
            .iter_items()
            .filter_map(|item| self.lower_interface_item(item))
            .collect();

        Some(hir::Interface { name, docs, items })
    }

    fn lower_interface_item(
        &mut self,
        item: ast::InterfaceItems<'_>,
    ) -> Option<hir::InterfaceItem> {
        if let Some(node) = item.func_item() {
            self.lower_func_item(node).map(hir::InterfaceItem::Func)
        } else if let Some(node) = item.typedef_item() {
            self.lower_type_definition(node)
        } else if let Some(_node) = item.use_item() {
            todo!()
        } else {
            None
        }
    }

    fn lower_type_definition(&self, node: ast::TypedefItem<'_>) -> Option<hir::InterfaceItem> {
        if let Some(type_item) = node.type_item() {
            self.lower_type_item(type_item)
        } else {
            None
        }
    }

    fn lower_func_item(&self, node: ast::FuncItem<'_>) -> Option<hir::FuncItem> {
        let docs = node.docs(self.src);
        let name = node.identifier(self.src)?.into();

        let ty = node.ty()?;
        let params = self.lower_params(ty.params()?)?;
        let return_value = ty
            .result_opt()
            .and_then(|node| self.lower_return_type(node));

        Some(hir::FuncItem {
            docs,
            name,
            params,
            return_value,
        })
    }

    fn lower_params(&self, ty: ast::ParamList<'_>) -> Option<im::Vector<hir::Parameter>> {
        let mut params = Vector::new();

        for param in ty.iter_params() {
            let name = param.name()?.value(self.src).into();
            let ty = self.lower_type(param.ty()?)?;
            params.push_back(hir::Parameter { name, ty });
        }

        Some(params)
    }

    fn lower_return_type(&self, _node: ast::ResultList<'_>) -> Option<hir::ReturnValue> {
        todo!()
    }

    fn lower_type(&self, ty: ast::Ty<'_>) -> Option<hir::Type> {
        if let Some(builtin) = ty.builtins() {
            self.lower_builtin(builtin)
        } else if let Some(handle) = ty.handle() {
            self.lower_handle(handle)
        } else if let Some(list) = ty.list() {
            self.lower_list(list)
        } else if let Some(option) = ty.option() {
            self.lower_option(option)
        } else if let Some(result) = ty.result() {
            self.lower_result(result)
        } else if let Some(tuple) = ty.tuple() {
            self.lower_tuple(tuple)
        } else if let Some(user_defined_type) = ty.user_defined_type() {
            self.lower_user_defined_type(user_defined_type)
        } else {
            None
        }
    }

    fn lower_builtin(&self, builtin: ast::Builtins<'_>) -> Option<hir::Type> {
        let name = builtin.value(self.src);

        match name {
            "u8" => Some(hir::Type::Builtin(hir::Builtin::U8)),
            "u16" => Some(hir::Type::Builtin(hir::Builtin::U16)),
            "u32" => Some(hir::Type::Builtin(hir::Builtin::U32)),
            "u64" => Some(hir::Type::Builtin(hir::Builtin::U64)),
            "s8" => Some(hir::Type::Builtin(hir::Builtin::I8)),
            "s16" => Some(hir::Type::Builtin(hir::Builtin::I16)),
            "s32" => Some(hir::Type::Builtin(hir::Builtin::I32)),
            "s64" => Some(hir::Type::Builtin(hir::Builtin::I64)),
            "float32" => Some(hir::Type::Builtin(hir::Builtin::Float32)),
            "float64" => Some(hir::Type::Builtin(hir::Builtin::Float64)),
            "char" => Some(hir::Type::Builtin(hir::Builtin::Char)),
            "bool" => Some(hir::Type::Builtin(hir::Builtin::Boolean)),
            "string" => Some(hir::Type::Builtin(hir::Builtin::String)),
            other => {
                unreachable!(
                    "Unknown builtin type, \"{other}\" at {}",
                    builtin.syntax().to_sexp()
                )
            }
        }
    }

    fn lower_handle(&self, handle: ast::Handle<'_>) -> Option<hir::Type> {
        if let Some(borrowed) = handle.borrowed_handle() {
            let _name = borrowed.name()?.value(self.src);
            todo!()
        } else if let Some(owned) = handle.owned_handle() {
            let _name = owned.name()?.value(self.src);
            todo!()
        } else {
            None
        }
    }

    fn lower_list(&self, list: ast::List<'_>) -> Option<hir::Type> {
        let element_type = list.ty().and_then(|ty| self.lower_type(ty))?;
        Some(hir::Type::List(Box::new(element_type)))
    }

    fn lower_option(&self, option: ast::Option_<'_>) -> Option<hir::Type> {
        let element_type = option.ty().and_then(|ty| self.lower_type(ty))?;
        Some(hir::Type::Option(Box::new(element_type)))
    }

    fn lower_result(&self, result: ast::Result_<'_>) -> Option<hir::Type> {
        let ok = result
            .ok_opt()
            .and_then(|ty| self.lower_type(ty))
            .map(Box::new);
        let err = result
            .err_opt()
            .and_then(|ty| self.lower_type(ty))
            .map(Box::new);
        Some(hir::Type::Result { ok, err })
    }

    fn lower_tuple(&self, tuple: ast::Tuple<'_>) -> Option<hir::Type> {
        let mut element_types = Vector::new();

        for ty in tuple.iter_tys() {
            element_types.push_back(self.lower_type(ty)?);
        }

        Some(hir::Type::Tuple(element_types))
    }

    fn lower_user_defined_type(
        &self,
        user_defined_type: ast::UserDefinedType<'_>,
    ) -> Option<hir::Type> {
        let text = user_defined_type.value(self.src);
        eprintln!("{:?} => {}", text, user_defined_type.syntax().to_sexp());
        todo!()
    }

    fn lower_type_item(&self, type_item: ast::TypeItem<'_>) -> Option<hir::InterfaceItem> {
        let docs = type_item.docs(self.src);
        let name = type_item.identifier(self.src)?.into();
        let ty = self.lower_type(type_item.ty()?)?;

        Some(hir::InterfaceItem::TypeAlias(hir::TypeAlias {
            docs,
            name,
            ty,
        }))
    }

    fn lower_package_decl(&self, root: ast::SourceFile<'_>) -> Option<hir::PackageDeclaration> {
        let node = root.package_opt()?;
        let docs = node.docs(self.src);

        let package_name = node.fully_qualified_package_name()?;
        let package = package_name.package()?.identifier()?.value(self.src).into();
        let path = package_name
            .path()?
            .iter_identifiers()
            .map(|p| p.value(self.src).into())
            .collect();
        let version = package_name.version_opt().map(|s| s.value(self.src).into());

        Some(hir::PackageDeclaration {
            docs,
            package,
            path,
            version,
        })
    }
}

fn named_item(item: ast::TopLevelItem<'_>) -> Option<NamedItem<'_>> {
    let node = item.syntax();

    if node.has_error() {
        return None;
    }

    if let Some(interface) = item.interface_item() {
        Some(NamedItem::Interface(interface))
    } else if let Some(world) = item.world_item() {
        Some(NamedItem::World(world))
    } else if let Some(import) = item.top_level_use_item() {
        Some(NamedItem::Use(import))
    } else {
        unreachable!("Unknown top level item node: {}", node.kind());
    }
}

#[derive(Debug, Copy, Clone)]
enum NamedItem<'a> {
    World(ast::WorldItem<'a>),
    Interface(ast::InterfaceItem<'a>),
    Use(ast::TopLevelUseItem<'a>),
}

impl<'a> NamedItem<'a> {
    fn node(self) -> tree_sitter::Node<'a> {
        match self {
            NamedItem::World(world) => world.syntax(),
            NamedItem::Interface(interface) => interface.syntax(),
            NamedItem::Use(node) => node.syntax(),
        }
    }

    fn ident(self) -> ast::Identifier<'a> {
        match self {
            NamedItem::World(w) => w.name().expect("syntax error"),
            NamedItem::Interface(i) => i.name().expect("syntax error"),
            NamedItem::Use(u) => {
                if let Some(ident) = u.alias_opt() {
                    return ident;
                }

                let use_path = u.use_path().expect("syntax error");
                if let Some(p) = use_path.fully_qualified_use_path() {
                    p.package()
                        .and_then(|pkg| pkg.identifier())
                        .expect("syntax error")
                } else if let Some(p) = use_path.local_use_path() {
                    p.identifier().expect("syntax error")
                } else {
                    unreachable!("syntax error")
                }
            }
        }
    }
}

impl<'a> From<ast::WorldItem<'a>> for NamedItem<'a> {
    fn from(value: ast::WorldItem<'a>) -> Self {
        NamedItem::World(value)
    }
}

impl<'a> From<ast::InterfaceItem<'a>> for NamedItem<'a> {
    fn from(value: ast::InterfaceItem<'a>) -> Self {
        NamedItem::Interface(value)
    }
}

#[cfg(test)]
mod tests {
    use crate::Compiler;

    use super::*;

    macro_rules! lowering_tests {
        (
            $(
                $( #[$meta:meta] )*
                $name:ident : $contents:literal
            ),* $(,)?
        ) => {
            $(
                #[test]
                #[allow(unused_mut, unused_variables)]
                $( #[$meta] )*
                fn $name() {
                    let db = Compiler::default();

                    let file = SourceFile::new(
                        &db,
                        format!("{}.wit", stringify!($name)).into(),
                        $contents.into(),
                    );
                    let ws = Workspace::new(&db, [(file.path(&db), file)].into_iter().collect());

                    let ast = crate::queries::parse(&db, file);
                    eprintln!("{}", ast.root_node(&db).to_sexp());

                    let got = super::lower(&db, ws, file);
                    let diags = super::lower::accumulated::<Diagnostics>(&db, ws, file);

                    assert!(diags.is_empty(), "{diags:#?}");

                    let mut settings = insta::Settings::clone_current();

                    #[derive(serde::Serialize)]
                    struct Info<'a> {
                        src: &'a str,
                        ast: String,
                    }
                    settings.set_info(&Info {
                        src: $contents,
                        ast: ast.root_node(&db).to_sexp(),
                    });
                    settings.set_omit_expression(true);

                    settings.bind(|| insta::assert_debug_snapshot!(got));
                }
            )*
        };
    }

    macro_rules! lowering_error_tests {
        (
            $(
                $( #[$meta:meta] )*
                $name:ident : $contents:literal
            ),* $(,)?
        ) => {
            $(
                #[test]
                #[allow(unused_mut, unused_variables)]
                $( #[$meta] )*
                fn $name() {
                    let db = Compiler::default();

                    let file = SourceFile::new(
                        &db,
                        format!("{}.wit", stringify!($name)).into(),
                        $contents.into(),
                    );
                    let ws = Workspace::new(&db, [(file.path(&db), file)].into_iter().collect());

                    let ast = crate::queries::parse(&db, file);
                    let diags = super::lower::accumulated::<Diagnostics>(&db, ws, file);

                    assert_ne!(diags.len(), 0, "No diagnostics emitted");

                    #[derive(serde::Serialize)]
                    struct Info<'a> {
                        src: &'a str,
                        ast: String,
                    }

                    let mut settings = insta::Settings::clone_current();
                    settings.set_info(&Info {
                        src: $contents,
                        ast: ast.root_node(&db).to_sexp(),
                    });
                    settings.set_omit_expression(true);

                    settings.bind(|| insta::assert_debug_snapshot!(diags));
                }
            )*
        };
    }

    lowering_tests! {
        lower_an_empty_file: "",
        lower_package_with_docs: "/// This is a package.\npackage wasi:filesystem@1.2.3;",
        empty_interface: "interface empty {}" ,
        interface_with_func: "interface console { log: func(message: string); }",
        interface_with_builtin_type: "interface i { type x = u32; }",
        empty_tuple: "interface i { type x = tuple<>; }",
        tuple_with_single_element: "interface i { type x = tuple<string>; }",
        tuple_with_multiple_elements: "interface i { type x = tuple<string, bool, u32>; }",
        result_with_ok_and_error: "interface i { type x = result<bool, string>; }",
        result_with_empty_ok: "interface i { type x = result<_>; }",
        bare_result: "interface i { type x = result; }",
        result_with_just_error: "interface i { type x = result<_, string>; }",
        list: "interface i { type x = list<u32>; }",
        option: "interface i { type x = option<u32>; }",

        empty_world: "world empty {}",
        #[ignore]
        world_with_function_export: "world console { export run: func(); }",
        #[ignore]
        world_with_external_export: "world with-import {
            export wasi:filesystem/filesystem;
        }",
        #[ignore]
        world_with_external_import: "world with-import {
            import wasi:filesystem/filesystem;
        }"
    }

    lowering_error_tests! {
        syntax_errors_are_emitted: "#$",
        #[ignore]
        refer_to_unknown_type: "interface i { type x = this-does-not-exist; }",
        #[ignore]
        recursive_records_are_not_allowed: "interface i { record recursive { inner: recursive } }",
        #[ignore]
        reference_cycles_are_not_allowed: "interface i {
            record first { second: second }
            record second { first: first }
        }",
        #[ignore]
        duplicate_identifiers: "interface i { record foo {} variant foo {} }",
    }
}
