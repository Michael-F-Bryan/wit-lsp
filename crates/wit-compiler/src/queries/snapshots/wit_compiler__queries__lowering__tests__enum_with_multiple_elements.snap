---
source: crates/wit-compiler/src/queries/lowering.rs
info:
  src: "interface i {\n            /// This is an enum.\n            enum foo {\n                /// first case.\n                first,\n                /// Second case.\n                /// and another\n                /// doc-comment.\n                second,\n             }\n         }"
  ast: "(source_file (top_level_item (interface_item name: (identifier) items: (interface_items (typedef_item (enum_item attributes: (attribute (doc_comment (docs))) name: (identifier) cases: (enum_case attributes: (attribute (doc_comment (docs))) name: (identifier)) cases: (enum_case attributes: (attribute (doc_comment (docs))) attributes: (attribute (doc_comment (docs))) attributes: (attribute (doc_comment (docs))) name: (identifier))))))))"
---
Package {
    decl: None,
    worlds: {},
    interfaces: {
        InterfaceIndex(0): Interface {
            name: i,
            docs: None,
            items: [
                Enum(
                    Enum {
                        name: foo,
                        index: EnumIndex(0),
                        docs: Some(
                            This is an enum.,
                        ),
                        cases: [
                            EnumCase {
                                name: first,
                                docs: Some(
                                    first case.,
                                ),
                            },
                            EnumCase {
                                name: second,
                                docs: Some(
                                    Second case.
                                    and another
                                    doc-comment.,
                                ),
                            },
                        ],
                    },
                ),
            ],
        },
    },
}
